### PR TITLE
Silent install auto run added

### LINKED ISSUE #4: Make launcher auto-install to library directory
Have the installer silent run, then make sure the selected client is the newly installed one.

### LINKED ISSUE #9: Download button doesn't work properly if installs folder isn't set
Repro:
- Delete all launcher JSONs.
- Run launcher.
- Press "download" button.
- You are prompted to select a directory.
- Select directory.
- "download" button is still grey as if downloading but no download is happening (log messages aren't being output to command window). Clicking "download" button asks if you want to cancel the download.

Or maybe I didn't wait long enough for the download to start??

### PR DESCRIPTION
**Pull request by two-one-five**
Dialog shows on finish, button disables while installing. Currently does not switch interface to select but since it all goes to the same folder anyway... the new install should overwrite the old one. Therefore the same interface should be selected anyway, given newbie user without multiple installs on purpose. Therefore it stands to reason that any newbie will get the intended functionality, but I have not tested to make sure that it works that way yet.

### CODE REVIEW

**Comment by ctrlaltdavid** [comment]
Downloaded installer doesn't run.
```
getSetting Return Value: C:\Users\UserName\AppData\Roaming\AthenaLauncher\storage
Installing silently, params: C:\Users\UserName\AppData\Roaming\AthenaLauncher\storage/Athena_Setup_Latest.exe C:\Users\UserName\AppData\Roaming\AthenaLauncher\storage\Athena_Interface_Latest_SILENT [
  '/S',
  '/D=C:\\Users\\UserName\\AppData\\Roaming\\AthenaLauncher\\storage\\Athena_Interface_Latest_SILENT'
]
Error: Command failed: C:\Users\UserName\AppData\Roaming\AthenaLauncher\storage/Athena_Setup_Latest.exe /S /D=C:\Users\UserName\AppData\Roaming\AthenaLauncher\storage\Athena_Interface_Latest_SILENT

    at ChildProcess.exithandler (child_process.js:308:12)
    at ChildProcess.emit (events.js:203:13)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5) {
  killed: false,
  code: 2,
  signal: null,
  cmd: 'C:\\Users\\UserName\\AppData\\Roaming\\AthenaLauncher\\storage/Athena_Setup_Latest.exe /S /D=C:\\Users\\UserName\\AppData\\Roaming\\AthenaLauncher\\storage\\Athena_Interface_Latest_SILENT'
}
```

This happens even if the "Athena_Interface_Latest_SILENT" directory exists.
Confirmed that Athena_Setup_Latest.exe exists.

Repro:
- Developer Command Prompt for VS2019 window; normal privileges (i.e., NOT run as administrator).
- Build the launcher, "Athena Launcher.exe".
- Run the launcher from command window (so that can see the log output in the window).
- Say "yes" to Windows prompt, "do you want this application from an unknown publisher to make changes to your device".
- Press the "download" button and once download is complete, the above error is encountered.
- A message box saying "installation complete" is displayed, which is also a bug - it should say "error".

UPDATE: Similar problem still occurs when trying to run the downloaded .exe. This is because Windows is preventing it from running. (When you try to run the downloaded .exe directly you get the Windows dialog, "Windows protected your PC" with "Don't run" button.

File: src/components/Dialogs/InstallComplete.vue [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<!--
+//  InstallComplete.vue
+//
+//  Created by Kalila L. on 15 Dec 2019.
```
Comment (ctrlaltdavid) [inline src/components/Dialogs/InstallComplete.vue]: Date wrong.

File: src/components/Dialogs/InstallComplete.vue [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+</template>
+
+
+<script>
+export default {
+  name: 'showInstallComplete',
+
+  data: () => ({
+      showInstallComplete: true,
+  }),
```
Comment (ctrlaltdavid) [inline src/components/Dialogs/InstallComplete.vue]: Indent needs 2 extra spaces.

File: src/components/Settings.vue [hunk 3]
```diff
@@ -282,7 +282,7 @@ export default {
             ipcRenderer.send('set-library-folder-default');
         },
         getLibraryFolder: function() {
-			ipcRenderer.send('getLibraryFolder');
+			ipcRenderer.send('get-library-folder');
```
Comment (ctrlaltdavid) [inline src/components/Settings.vue]: Tabs.

**Comment by ctrlaltdavid** [comment]
Downloading Athena_Setup_Latest.exe again doesn't overwrite already existing copy of the file. You can see this by the date stamp of the file, and also if you replace the original with another file with the sane name.

**Comment by two-one-five** [comment]
> 
> 
> Downloading Athena_Setup_Latest.exe again doesn't overwrite already existing copy of the file. You can see this by the date stamp of the file, and also if you replace the original with another file with the sane name.

Is it possible that because the actual file (hash, for example) did not change that it is not overwriting it? What if we put in a completely unrelated exe and renamed it to that and then had it download?

**Comment by ctrlaltdavid** [comment]
> Is it possible that because the actual file (hash, for example) did not change that it is not overwriting it? What if we put in a completely unrelated exe and renamed it to that and then had it download?

I checked this but is really isn't replacing the file; "also if you replace the original with another file with the sa[m]e name". … Also tested by renaming and hex editing the installer .exe and downloading from a test server. The pre-existing Athena_Setup_Latest.exe is not replaced.

So probably need to update background.js to delete the target .exe if it already exists before starting the download. (Electron-dl appears not to have any "overwrite" option.)

File: src/background.js [hunk 4]
```diff
@@ -550,46 +550,90 @@ function launchInstaller() {
...
+    var executablePath;
+    
+    getSetting('athena_interface.library', storagePath.default).then(function (libPath) {    
+        if (libPath) {
+            executablePath = libPath + "/Athena_Setup_Latest.exe";
+        } else {
+            executablePath = storagePath.default + "/Athena_Setup_Latest.exe";
+        }
+        
+        var installPath = libPath + "\\Athena_Interface_Latest_SILENT";
```
Comment (ctrlaltdavid) [inline src/background.js]: Per the preceding `if()`, `libPath` may not be set.

**Comment by ctrlaltdavid** [comment]
The grey "downloading" button changes back to blue "download" for a few seconds after downloading the installer before changing to grey "installing". (Perhaps when virus scanner is scanning the file?)

**Comment by ctrlaltdavid** [comment]
The error catching appears not to work in the case of Windows blocking the installer from running. The launcher stays with grey "installing, please wait" button state and the following is reported to the console:
```
getSetting Return Value: C:\Users\UserName\AppData\Roaming\AthenaLauncher\storage
Installing silently, params: C:\Users\UserName\AppData\Roaming\AthenaLauncher\storage/Athena_Setup_Latest.exe C:\Users\UserName\AppData\Roaming\AthenaLauncher\storage\Athena_Interface_Latest_SILENT [
  '/S',
  '/D=C:\\Users\\UserName\\AppData\\Roaming\\AthenaLauncher\\storage\\Athena_Interface_Latest_SILENT'
]
(node:14348) UnhandledPromiseRejectionWarning: Error: spawn EBUSY
    at ChildProcess.spawn (internal/child_process.js:394:11)
    at spawn (child_process.js:553:9)
    at execFile (child_process.js:237:17)
    at module.<computed> (electron/js2c/asar.js:163:31)
    at D:\Workspaces\CtrlAltStudio\Pantheon-Launcher\Pantheon-Launcher\dist_electron\win-ia32-unpacked\resources\app.asar\background.js:21:16537
(node:14348) UnhandledPromiseRejectionWarning: Error: spawn EBUSY
    at ChildProcess.spawn (internal/child_process.js:394:11)
    at spawn (child_process.js:553:9)
    at execFile (child_process.js:237:17)
    at module.<computed> (electron/js2c/asar.js:163:31)
    at D:\Workspaces\CtrlAltStudio\Pantheon-Launcher\Pantheon-Launcher\dist_electron\win-ia32-unpacked\resources\app.asar\background.js:21:16537
(node:14348) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:14348) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:14348) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:14348) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

File: src/components/Dialogs/InstallFailed.vue [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        persistent
+        v-model="showInstallFailed"
+    >
+        <v-card>
+            <v-card-title
+                class="headline"
+                primary-title
+                dark
+            >
+                <v-icon color="success" class="mr-2">mdi-message-alert</v-icon>
```
Comment (ctrlaltdavid) [inline src/components/Dialogs/InstallFailed.vue]:
```suggestion
                <v-icon color="success" class="mr-2">mdi-message-alert</v-icon>
```
                <v-icon color="red" class="mr-2">mdi-message-alert</v-icon>

And similarly for other error dialogs.

Comment (ctrlaltdavid) [inline src/components/Dialogs/InstallFailed.vue]:
```suggestion
                <v-icon color="red" class="mr-2">mdi-message-alert</v-icon>
```

**Comment by ctrlaltdavid** [comment]
Installation of the downloaded installer fails if Sandbox is running. It just reports a message gox, "Installation Failed" with no explanation.

File: src/background.js [hunk 6]
```diff
@@ -551,11 +550,25 @@ function launchInstaller() {
...
     var exeLocToInstall; // This is what gets installed.
+    var list = await tasklist();
+    var canInstall = true;
+    
+    list.forEach((task) => {
+        if (task.imageName == "server-console.exe") {
+            console.log("SANDBOX FOUND!");
+            canInstall = false;
+        }
+    })
```
Comment (ctrlaltdavid) [inline src/background.js]:
```suggestion
    });
```

File: src/plugins/store.js [hunk 7]
```diff
@@ -23,6 +23,7 @@ export const store = new Vuex.Store({
 		populatedInterfaceList: [],
 		noSteamVR: false,
 		currentLibraryFolder: "No library folder selected.",
+        currentError: ""
```
Comment (ctrlaltdavid) [inline src/plugins/store.js:26]: Indenting suspect.

File: src/background.js [hunk 8]
```diff
@@ -551,11 +550,25 @@ function launchInstaller() {
...
+    
+    if (!canInstall) {
+        win.webContents.send('silent-installer-failed', 'Your server sandbox is running, please close it before proceeding.');
```
Comment (ctrlaltdavid) [inline src/background.js]:
```suggestion
        win.webContents.send('silent-installer-failed', 'Your server sandbox is running. Please close it before proceeding.');        
```

File: src/background.js [hunk 9]
```diff
@@ -551,11 +550,25 @@ function launchInstaller() {
...
+async function silentInstall() {
     var executableLocation; // This is the downloaded installer.
     var installPath; // This is the location to install the application to.
     var executablePath; // This is the location that the installer exe is located in after being downloaded.
     var exeLocToInstall; // This is what gets installed.
+    var list = await tasklist();
+    var canInstall = true;
+    
+    list.forEach((task) => {
+        if (task.imageName == "server-console.exe") {
```
Comment (ctrlaltdavid) [inline src/background.js]:
```suggestion
        if (task.imageName === "server-console.exe") {
```

**Review by ctrlaltdavid — APPROVED** [review]

File: src/plugins/store.js (hunk 7 above)
Comment (two-one-five) [reply src/plugins/store.js:26]: That indenting was correct, it was all the surrounding code that was incorrect! It's been fixed. :)

**Pull request merged**